src/main/java/build/buildfarm/worker/shard/Worker.java

```diff
@@ -140,53 +134,48 @@ public class Worker {
   private Pipeline pipeline;
   private Backplane backplane;
   private LoadingCache<String, Instance> workerStubs;
-  @Autowired private AwsAdmin awsAdmin;
 
-  /**
-   * The method will prepare the worker for graceful shutdown and send out grpc request to disable
-   * scale in protection when the worker is ready. If unexpected errors happened, it will cancel the
-   * graceful shutdown progress make the worker available again.
-   */
+  /** The method will prepare the worker for graceful shutdown when the worker is ready. */
   public void prepareWorkerForGracefulShutdown() {
-    inGracefulShutdown = true;
-    log.log(
-        Level.INFO,
-        "The current worker will not be registered again and should be shutdown gracefully!");
-    pipeline.stopMatchingOperations();
-    int scanRate = 30; // check every 30 seconds
-    int timeWaited = 0;
-    int timeOut = 60 * 15; // 15 minutes
+    if (configs.getWorker().getGracefulShutdownSeconds() > 0) {
+      inGracefulShutdown = true;
+      System.err.println(
+          "Graceful Shutdown - The current worker will not be registered again and should be shutdown gracefully!");
+      pipeline.stopMatchingOperations();
+      int scanRate = 30; // check every 30 seconds
+      int timeWaited = 0;
+      int timeOut = configs.getWorker().getGracefulShutdownSeconds();
 
-    try {
-      while (!pipeline.isEmpty() && timeWaited < timeOut) {
-        SECONDS.sleep(scanRate);
-        timeWaited += scanRate;
-        log.log(INFO, String.format("Pipeline is still not empty after %d seconds.", timeWaited));
-      }
-    } catch (InterruptedException e) {
-      log.log(Level.SEVERE, "The worker gracefully shutdown is interrupted: " + e.getMessage());
-    } finally {
-      // make a grpc call to disable scale protection
-      String clusterEndpoint = configs.getServer().getAdmin().getClusterEndpoint();
-      log.log(
-          INFO,
-          String.format(
-              "It took the worker %d seconds to %s",
-              timeWaited,
-              pipeline.isEmpty() ? "finish all actions" : "but still cannot finish all actions"));
       try {
-        awsAdmin.disableHostScaleInProtection(clusterEndpoint, configs.getWorker().getPublicName());
-      } catch (Exception e) {
-        log.log(
-            SEVERE,
+        if (pipeline.isEmpty()) {
+          System.err.println("Graceful Shutdown - no work in the pipeline.");
```

Comment:
do you we prefer System.err.println over logging here? any specific reason?

Reply:
The logging does not work at this stage since it it looks like it stops logging before PreDestroy is called. So we output to stderr to match the shutdown message.

Reply:
sounds like this is something we want to fix or at least there should be some comments about why we have to use errr.println.